### PR TITLE
feat: [#427] Improve Mail Module with Interface-based Mailing and Make Command

### PR DESCRIPTION
**Pull request by kkumar-gcc**
## 📑 Description

Closes https://github.com/goravel/goravel/issues/427

## ✅ Checks

- [ ] Added test cases for my code

### CODE REVIEW

File: mail/console/stubs.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package console
+
+type Stubs struct {
+}
+
+func (s Stubs) Mail() string {
```
Comment (hwbrzzl) [inline mail/console/stubs.go:6]: Missing the `Queue` method?

File: mail/application_test.go [hunk 2]
```diff
@@ -109,7 +112,8 @@ func (s *ApplicationTestSuite) TestQueueMail() {
 		Cc([]string{testCc}).
 		Bcc([]string{testBcc}).
 		Attach([]string{"../logo.png"}).
-		Content(mail.Content{Subject: "Goravel Test Queue", Html: "<h1>Hello Goravel</h1>"}).
+		Subject("Goravel Test Queue").
+		Content(mail.Content{Html: "<h1>Hello Goravel</h1>"}).
 		Queue())
```
Comment (hwbrzzl) [inline mail/application_test.go:117]: Missing test cases for `Queue` with variable.

File: mail/application_test.go [hunk 3]
```diff
@@ -80,7 +82,8 @@ func (s *ApplicationTestSuite) TestSendMailWithFrom() {
 		Cc([]string{testCc}).
 		Bcc([]string{testBcc}).
 		Attach([]string{"../logo.png"}).
-		Content(mail.Content{Subject: "Goravel Test 587 With From", Html: "<h1>Hello Goravel</h1>"}).
+		Subject("Goravel Test 587 With From").
+		Content(mail.Content{Html: "<h1>Hello Goravel</h1>"}).
 		Send())
```
Comment (hwbrzzl) [inline mail/application_test.go:87]: Missing test cases for `Send` with variable.

File: mail/application.go [hunk 4]
```diff
@@ -111,7 +129,40 @@ func (r *Application) instance() *Application {
 	return r
 }
 
-func SendMail(config config.Config, subject, html string, fromAddress, fromName string, to, cc, bcc, attaches []string) error {
+func (r *Application) setUsingMailable(mailable mail.Mailable) {
+	if *mailable.Content() != (mail.Content{}) {
+		r.content = *mailable.Content()
+	}
```
Comment (hwbrzzl) [inline mail/application.go:135]: Why is there sucha judgment?

File: contracts/mail/mail.go [hunk 5]
```diff
@@ -13,15 +13,31 @@ type Mail interface {
...
+type Mailable interface {
+	// Envelope set the envelope of Mailable.
+	Envelope() *Envelope
+	// Content set the content of Mailable.
+	Content() *Content
+	// Attachments set the attachments of Mailable.
+	Attachments() []string
+}
+
+type ShouldQueue interface {
```
Comment (hwbrzzl) [inline contracts/mail/mail.go:33]:
```suggestion
type QueueMailable interface {
```

File: contracts/mail/mail.go [hunk 6]
```diff
@@ -33,3 +49,14 @@ type From struct {
...
 }
+
+type Envelope struct {
+	From     From
+	To       []string
+	Cc       []string
+	Bcc      []string
+	ReplyTo  []string
+	Tags     []string
+	Metadata map[string]any
```
Comment (hwbrzzl) [inline contracts/mail/mail.go:60]: These fields seem not to be used.

**Review by hwbrzzl — COMMENTED** [review]
Great PR 👍

File: mail/console/stubs.go (hunk 1 above)
Comment (kkumar-gcc) [reply mail/console/stubs.go:6]: Queue method is only required when you need to queue the mail. So it's optional.

File: mail/application_test.go (hunk 2 above)
Comment (kkumar-gcc) [reply mail/application_test.go:117]: Yeah test cases are no finished yet.

File: mail/application_test.go [hunk 7]
```diff
@@ -171,3 +206,32 @@ func mockConfig(mailPort, redisPort int) *configmock.Config {
...
+	return &mail.Envelope{
+		Bcc:     []string{testBcc},
+		Cc:      []string{testCc},
+		From:    mail.From{Address: testFromAddress, Name: testFromName},
+		Subject: "Goravel Test 587 With Mailable",
+		To:      []string{testTo},
+	}
+}
+
+func (m *TestMailable) Queue() *mail.Queue {
```
Comment (hwbrzzl) [inline mail/application_test.go:235]: This function doesn't exist in `stubs.go`.

File: mail/application.go [hunk 8]
```diff
@@ -111,7 +129,43 @@ func (r *Application) instance() *Application {
 	return r
 }
 
-func SendMail(config config.Config, subject, html string, fromAddress, fromName string, to, cc, bcc, attaches []string) error {
+func (r *Application) setUsingMailable(mailable mail.Mailable) {
+	content := mailable.Content()
+	if content != nil {
+		if content.Html != "" {
```
Comment (hwbrzzl) [inline mail/application.go:135]:
```suggestion
	if content != nil && content.Html != "" {
```

File: contracts/mail/mail.go [hunk 9]
```diff
@@ -1,27 +1,43 @@
...
+type Mailable interface {
+	// Attachments set the attachments of Mailable.
+	Attachments() []string
+	// Content set the content of Mailable.
+	Content() *Content
+	// Envelope set the envelope of Mailable.
+	Envelope() *Envelope
+}
+
+type ShouldQueue interface {
```
Comment (hwbrzzl) [inline contracts/mail/mail.go:33]: Can we only use `Mailable`? Add the `Queue` function to `Mailable` directly.

**Review by devhaozi — APPROVED** [review]

**Pull request merged**